From 5a9c2ef0cf8947c452230b90e89f4e34b8d6f72d Mon Sep 17 00:00:00 2001
From: John Johansen <john.johansen@canonical.com>
Date: Fri, 18 Mar 2016 06:05:47 -0700
Subject: [PATCH 1616/1767] Revert "UBUNTU: SAUCE: apparmor: Fix: deadlock in
 aa_put_label() call chain"

BugLink: http://bugs.launchpad.net/bugs/1379535

This reverts commit 981d09c847dd7a382c069e81b4dd9ae0b180d698.

Signed-off-by: Tim Gardner <tim.gardner@canonical.com>
---
 security/apparmor/label.c | 16 ++++++----------
 1 file changed, 6 insertions(+), 10 deletions(-)

diff --git a/security/apparmor/label.c b/security/apparmor/label.c
index 4c7fb19..627109a 100644
--- a/security/apparmor/label.c
+++ b/security/apparmor/label.c
@@ -234,8 +234,10 @@ void aa_label_free(struct aa_label *label)
 	kfree(label);
 }
 
-static void label_free_switch(struct aa_label *l)
+static void label_free_rcu(struct rcu_head *head)
 {
+	struct aa_label *l = container_of(head, struct aa_label, rcu);
+
 	if (l->flags & FLAG_NS_COUNT)
 		aa_free_namespace(labels_ns(l));
 	else if (label_isprofile(l))
@@ -244,14 +246,6 @@ static void label_free_switch(struct aa_label *l)
 		aa_label_free(l);
 }
 
-static void label_free_rcu(struct rcu_head *head)
-{
-	struct aa_label *l = container_of(head, struct aa_label, rcu);
-
-	(void) aa_label_remove(labels_set(l), l);
-	label_free_switch(l);
-}
-
 bool aa_label_remove(struct aa_labelset *ls, struct aa_label *label);
 void aa_label_kref(struct kref *kref)
 {
@@ -260,10 +254,12 @@ void aa_label_kref(struct kref *kref)
 
 	if (!ns) {
 		/* never live, no rcu callback needed, just using the fn */
-		label_free_switch(l);
+		label_free_rcu(&l->rcu);
 		return;
 	}
 
+	(void) aa_label_remove(&ns->labels, l);
+
 	/* TODO: if compound label and not invalid add to reclaim cache */
 	call_rcu(&l->rcu, label_free_rcu);
 }
-- 
1.9.3

